### DISCUSSION TITLE
Role overlapping with Customer, Teams & Projects

### DISCUSSION THREAD

**Discussion by xstable** [kimai/kimai#4165] [Questions & Answers]
I have:
-  a Customer "Company1".
- two teams: "Team-1" and "Team2"
- two projects for this Company1: `Proj1` & `Proj2` 

## Team-1 Settings:
- 8 Users, 2 of it are teamleads (let's name it `TeamL1` and `TeamL2`)
- `TeamL1` is my admin-user, `TeamL2` is a general team-member with role `Teamleader` (to monitor reportings for the whole teams)
- `Grant access to customers` only allowed to `Company1`
- `Grant access to projects` only set to `Proj1`

## Team-2 Settings:
- 3 Users, 1 Teamlead (**BUT it's not `TeamL2` from above!!**)
-  `Grant access to customers` only allowed to `Company1`
- `Grant access to projects` only set to `Proj2`

> **IMPORTANT**
> There is one Teammate which assigned in both teams (let's name him "JIM")

## ISSUE
If `TeamL2` login to his account and look at the `Monthly view for all user`, he see `JIM` in his list.
*Generally that is ok, as `JIM` is assigned to both teams. To be clear, JIM didn't do any work in Proj1 for a long time, but worked ad Proj2 in the last month)*
But if I click on `JIM's` wonking-hours to see his details, screen witch into `Monthly view for one user` and I see: 
1. `Company1` as headline, followed by the `Proj2` and his Activities (he has only Activity `Coding`)

As TeamL2 is only Teamleader with Permissions to `Team-1` he shouldn't see `Proj2` in this listing, as this Project is only available for `Team-2` where `TeamL2` isn't a part of.

Did I setup the Team's in a wrong way?

####
**BTW:**
I've tried now to remove `company1` from `Team1`, but only keep `Proj1` for this team. 
Same result, TeamL2 still see `Proj2` (even after re-login)

**Comment by xstable** [comment]
# Another strange issue with that:

As I've described above: `TeamL1` of `Proj1` is me (admin-role) and I'm also Teamlead on this `Proj1`.
If `TeamL2` open `Projects by month, activity and user` he see even other Projects (e.g `Proj3`) I've part of. That Project is completly outside of any `Project` or `Customer` he is assigned to.

Might the issue be, that if a user has the role `TeamLead` in any project, that he can (accidentally) see all projects his Team-Mates are in?

*As this is a pretty complex setup, I'm ready for a screen-sharing session, if that simplify reproducing this issue*

**Comment by xstable** [comment]
Can somebody reproduce and - agree or disagree - with this issue?

I can't currently use kimai in my project if this bug exists. As I have teams that shouldn't see other project-content at all.

**Reply by kevinpapst** [reply]
Teamleads can see times of their team members.
Those times are not limited, because the teamlead needs to be able to understand what his team-mate was working on.
Even if that was for a different team. 

The same behavior exists for every user who can see all of his own records. 
Even if he cannot book times on the project anymore, as permissions were removed. 
But still all entries from the past are visible.

This behavior is on purpose.
Edit: and it will not be changed.

**Reply by suther** [reply]
OK, I see. 
Do you have a hint, how to setup a new Role for an ’project based’ Teamlead?
Means that teamleader can only see times of the current project of all teammembers?

**Reply by kevinpapst** [reply]
Only possible if each user is in exactly one team.

**Reply by xstable** [reply]
> Teamleads can see times of their team members. Those times are not limited, because the teamlead needs to be able to understand what his team-mate was working on. Even if that was for a different team.
> ...
> This behavior is on purpose. Edit: and it will not be changed.

I fully understand your point, and it's a **great feature!**

Maybe we can implement an **additional** option to cover both cases.
In my experience, many "decentrally organized" remote IT agencies have the following requirements for their team structure (as - for Reporting - they need a project-based approach, and not a team-based)

- Think about a de-centralized company which have 5 teamleader and 20 employees / freelancer.
- Teamleader _can only see Projects he is assigned to_. 
**teamleader should never see what other projects the "company" or their teammates are working on. (for security reasons)**
- Some of the Developers are in more than one team (which might have different teamleaders).

The perfect solution would be, if we could add an option to Kimai to choose between following settings for teamleader-role: 
1. Teamleader see any projects where their teammates are assigned to (_as it's currently is_)
2. Teamleader see only projects, where the teamleader and his teammates are assigned, 

Sure, teamleader wouldn't get the total (project overlaping) hourly overview of his team mate, 
BUT: 
1. in this way of workflow it isn't needed anyway, and
2. if needed, team-mates can tell any team-leader of their participating projects how much hours per month they are able to work for him. 

As far as I see, this option would only be needed for 4 Reporting-Sections. Those who do reporting for a sigle user: 
![image](https://github.com/kimai/kimai/assets/1778068/3a2f0276-83f9-46d2-83ae-ed5f3154b56a)

What do you think about this @kevinpapst
